Kubeone addons API (#779)
kubermatic/docs: content/kubeone/master/guides/addons/_index.en.md

> 

Reviewer: I think we should get rid of the link to the design proposal on L14 because it's heavily outdated. There's an explanation about how addons work (including `--prune` and everything else) at the bottom of this document.

Author: 👍

**Note:** The `b64enc` function is a [`sprig` function][sprig-b64enc].

## Addons API Reference

> Take a look at the [KubeOne's autogenerated addons API][addons-api].

Reviewer: Would it look better if we would expand this sentence:

```suggestion
Take a look at the [KubeOne's autogenerated API reference][addons-api] for more details about available options in the Addons API.
```

> 

Reviewer: This looks a little bit weird when placed here. I think it might look better if we move this section to the beginning (before `Writing Addons`).

Author: The idea is to point to the API reference before demonstrating YAMLs that are immediately follow in the next "Enabling Addons" section.

apiVersion: kubeone.io/v1beta1
kind: KubeOneCluster
versions:

>   kubernetes: 1.20.1

Reviewer: We can already put `1.22.1` here. :slightly_smiling_face:

Author: but 1.20.1 is still supported ;)

{{% notice note %}}
Subdirectories are not considered when applying addons. Only addons in the root
of the provided directory will be applied. E.g. `./addons/addon_name1` is
supported addon while `./addons/groupping/addon_name1` and

> `./addons/addon_name1/subdir1` does not.

Reviewer: This is a bit similar to the previous notice, so it might be confusing. What do you think about something like this:

> Addons can be organized into subdirectories, but only one level of subdirectories is supported. For example, `./addons/example-addon-1` is supported and YAML manifest in that directory will be deployed, but `./addons/example-addon-1/subdirectory-2` directory will be entirely ignored.

{{% /notice %}}

## Embedded Addons

> Since v1.3 release KubeOne has new kind of addons that always caried with the

Reviewer: I'd shuffle words a little bit. Also, some typo and grammar fixes.

```suggestion
Since KubeOne v1.3 release, there's a new kind of addons that are always carried with the
```

> binariy itself, embedded addons. This is done thanks to new Go's

Reviewer: I think an em dash would look better than a comma. Also, some typos.

```suggestion
binary itself — embedded addons. This is done thanks to the new Go's
```

[`embed`][embed-docs] package. 

Some of those addons are directly and automatically used by the KubeOne itself

> (like machinecontroller for example). But it's not the case for all of them.

Reviewer: ```suggestion
(for example `machine-controller`), but it's not the case for all of them.
```

Author: the addon is machinecontroller, not machine-controller

> Some very useful addons need to be activated by the user in order to be deployed

Reviewer: ```suggestion
Some very useful addons can be activated by the user to be deployed
```

> to the cluster, like `backups-restic`, `cluster-autoscaler` or

Reviewer: WDYT about also providing links to those addons?

> `unattended-upgrades`.

Reviewer: I'd also add `default-storage-class` because that's basically only 4 addons we currently have.

> ### Activate Embedded Eddons

Reviewer: One problem here that we might encounter is that some addons might require some parameters to be passed. For example, `backups-restic` require `.Params.resticPassword`, `.Params.s3Bucket`, and `.Params.awsDefaultRegion` to be provided. We should add a note block about this and maintain the list of required parameters somewhere.

> Or order to activate such embedded addons, user need to use new Addons
> [API][addons-api].

Reviewer: I'd extend the hyperlink to make it easier to click on it (API is a short word). Also, some typos and improvements.

```suggestion
To activate the embedded addons, the user needs to use the new [Addons API][addons-api].
```

> 

Reviewer: We should add a warning box to this section to highlight that the addons directory must exist even if you don't plan on providing your custom addons. Something along:

```
{{% notice warning %}}
The addons directory must exist even if you plan to only use embedded addons without providing your custom addons. This is a known bug and can be tracked on [GitHub](https://github.com/kubermatic/kubeone/issues/1496).
{{% /notice %}}
```

Author: 👍

> Example:

Reviewer: ```suggestion
Example:

```

Add a new line to make markdown linters happy. :slightly_smiling_face:

Author: 👍

```yaml
apiVersion: kubeone.io/v1beta1
kind: KubeOneCluster
versions:
  kubernetes: 1.20.1

addons:
  enable: true
  path: "./addons"
  addons:
  - name: cluster-autoscaler
  - name: unattended-upgrades
```

> [Up to date (and possibly unreleased yet) list of addons.][addons-list-url]

Reviewer: I'd remove `(and possibly unreleased yet)` because we might very easily forget to remove this when releasing the stable version of docs. Someone browsing the master version of the docs should be aware that many things might be unreleased. Moreover, there's a warning at the top:

> This version is under construction, please use an official release version

```suggestion
[Up to date list of addons is available on GitHub.][addons-list-url]
```

Reviewer: Additionally, users might not know which addons are deployed by default and which are provided to be used with the Addons API. We should maintain a complete list somewhere, probably both on GitHub and docs.

Author: The phase `(and possibly unreleased yet)` is a look past 1.3 release, in the future. When we will add some new addons in the future, there will a time gap in between a releases which falls right under "not released yet". This phase is not be to deleted after the release.

### Overriding Embedded Eddons

> Some of those embedded addons ad used to be written in Go structures form. Now

Reviewer: ```suggestion
Some of those embedded addons used to be written in Go structures form. Now
```

KubeOne has them in YAML form so those addons are easier to maintain and to let

> users to overrider them.

Reviewer: ```suggestion
users to override them.
```

> For example if you wish to have own machinecontroller manifests being deployed
> to the cluster, you cloud override `machinecontroller` addons. Enable addons
> like described above and in the addons directory place your machine-controller
> manifests in the same directory name, e.g.
> `./addons/machinecontroller/yourmanifest.yaml`.

Reviewer: ```suggestion
For example, if you wish to have your own machine-controller manifests being deployed
to the cluster, you can override the `machinecontroller` addon. Enable addons
like described above and in the addons directory place your `machinecontroller`
manifests in the same directory name, e.g.
`./addons/machinecontroller/yourmanifest.yaml`.
```

## Cleanup Addons

There might be a situation when user might want to remove the addon

> declaratively. You will need to use new `delete` filed from Addons [API].

Reviewer: ```suggestion
declaratively. You can use the new `delete` field from the [Addons API].
```

```yaml
apiVersion: kubeone.io/v1beta1
kind: KubeOneCluster
versions:
  kubernetes: 1.20.1

addons:
  enable: true
  path: "./addons"
  addons:
  - name: unattended-upgrades
    delete: true
```

{{% notice note %}}
That will cleanup all the traced of `unattended-upgrades` addon, but only it's
manifests, not the contents that addons workloads might have produced, i.e.
backups themselves.

> {{% /notice %}}

Reviewer: I'd also add a warning that `delete` only works for embedded addons, at least for now.

## Parameters

It's possible to pass down user defined parameters (in key/value form) down to

> an addon. There are two fields (of type map[string]string):

Reviewer: ```suggestion
an addon. There are two fields (of type `map[string]string`):
```

> `.addons.globalParams` and `.addons.addons[] .params` in each addon.

Reviewer: It might look better with bullet-points. I'd also add a note such as this one.

```suggestion

* `.addons.globalParams`
* `.addons.addons[].params` in each addon.

{{% notice note %}}
`.addons.addons[].params` has higher priority over `.addons.globalParams`, so you can use it to override globally defined parameters.
{{% /notice %}}
```
